OpenRA.Mods.Common/Traits/StoresResources.cs

```diff
@@ -9,62 +9,100 @@
  */
 #endregion
 
-using OpenRA.Primitives;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Adds capacity to a player's harvested resource limit.")]
-	public class StoresResourcesInfo : TraitInfo
+	public class StoresResourcesInfo : TraitInfo, IStoresResourcesInfo
 	{
 		[FieldLoader.Require]
-		public readonly int Capacity = 0;
+		[Desc("The amounts of resources that can be stored.")]
+		public readonly int Capacity = 28;
+
+		[Desc("Which resources can be stored.")]
+		public readonly string[] Resources = Array.Empty<string>();
+
+		string[] IStoresResourcesInfo.ResourceTypes => Resources;
 
 		public override object Create(ActorInitializer init) { return new StoresResources(init.Self, this); }
 	}
 
-	public class StoresResources : INotifyOwnerChanged, INotifyCapture, IStoreResources, ISync, INotifyKilled, INotifyAddedToWorld, INotifyRemovedFromWorld
+	public class StoresResources : IStoresResources, ISync
 	{
+		readonly Dictionary<string, int> contents = new();
 		readonly StoresResourcesInfo info;
-		PlayerResources player;
 
 		[Sync]
-		public int Stored => player.ResourceCapacity == 0 ? 0 : (int)((long)info.Capacity * player.Resources / player.ResourceCapacity);
+		public int ContentHash
+		{
+			get
+			{
+				var value = 0;
+				foreach (var c in contents)
+					value += c.Value << c.Key.Length;
+
+				return value;
+			}
+		}
+
+		public int ContentsSum { get; private set; } = 0;
```

Comment:
Expensive sync.

Reply:
The line above. Is it really worth while to sync the value of each harvester contineously. And you do this by looping over the contents. I guess there are only a few types of content.